tests/ttasks.nim

```diff
@@ -0,0 +1,42 @@
+import std/os
+
+import chronos
+import chronos/threadsync
+import chronos/unittest2/asynctests
+import taskpools
+
+import apatheia/queues
+import apatheia/tasks
+
+## todo: setup basic async + threadsignal + taskpools example here
+## 
+
+proc addNums(a, b: float): float {.asyncTask.} =
+  os.sleep(100)
+  echo "adding: ", a, " + ", b
+  return a + b
+
+proc addNumValues(vals: openArray[float]): float {.asyncTask.} =
+  os.sleep(100)
+  result = 0.0
+  for x in vals:
+    result += x
+  echo "adding sums: ", vals
+
+suite "async tests":
+  var tp = Taskpool.new(num_threads = 2) # Default to the number of hardware threads.
+
+  asyncTest "test addNums":
+    var jobs = newJobQueue[float](taskpool = tp)
+    echo "\nstart"
+    let res = await jobs.submit(addNums(1.0, 2.0,))
+    echo "result: ", res.repr
+    check true
+
+  asyncTest "test addNumValues":
+    var jobs = newJobQueue[float](taskpool = tp)
+    echo "\nstart"
+    let args = @[1.0, 2.0, 3.0]
+    let res = await jobs.submit(addNumValues(args))
```

Comment:
@tbekas this currently doesn't support back-pressure. However making `submit` async adds overhead and just changes the queue filling to Chronos's queue.

Instead, what do you think about an additional api `trySubmit(...): Option[Future[T]]` which would only submit and return the future if the queue isn't full. Then the upstream code could handle / reject processing jobs appropriately.

The other alternative would be throwing an exception when the queue is full, though that wouldn't fit the Codex codebase as well.

Reply:
@tbekas I'm going to merge this PR. Another can be opened for this.

Reply:
I like the `trySubmit(...): Option[Future[T]]` API 👌